cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -89,13 +100,117 @@ public MetadataEntity append(String key, String value) {
     return new MetadataEntity(existingParts);
   }
 
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    MetadataEntity that = (MetadataEntity) o;
+    return Objects.equals(details, that.details);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(details);
+  }
+
+  @Override
+  public String toString() {
+    return "MetadataEntity{" +
+      "details=" + details +
+      '}';
+  }
+
   /**
    * @return A {@link List} of {@link KeyValue} representing the metadata entity
    */
   public List<KeyValue> getKeyValues() {
     return details;
   }
 
+  /**
+   * @return A {@link Iterator} over the keys in this {@link MetadataEntity}
+   */
+  public Iterator<String> getKeys() {
+    return new MetadataEntityKeyIterator();
```

Comment:
If you return an `Iterable`, you can use Java stream to help, no need to create a new class.

```
  public Iterable<String> getKeys() {
    return details.stream().map(KeyValue::getKey)::iterator
  }
```

Reply:
And if you stick to returning `Iterator` (which I don't think you should), you can also use stream

```
public Iterator<String> getKeys() {
  return details.stream().map(KeyValue::getKey).iterator();
}
```